### PR TITLE
Deprecate built-in GraphQL support

### LINKED ISSUE #619: GraphQL support should be a third party application.
I'd like our graphql support to move out into being a third party package.

We should then document:

* The third party package it's been moved to.
* Links to ASGI implementations with ariadne, tartiflette, strawberry.

Questions:

* Which of Ariadne, Tartiflette & Strawberry currently have provide an ASGI application, that Starlette can mount?
* Are there any other GraphQL implementations I've missed here?
* Anyone keen to take on the GraphQL implementation as a third party package?

### PR DESCRIPTION
**Pull request by JayH5**
See #619

I'm trying to do some house-keeping so that we can better maintain Starlette. We need to either put some major work into the GraphQL integration (upgrade to Graphene 3, add a bunch of stuff people want), or deprecate it. I'm in favour of the latter, since there doesn't seem to be much interest in maintaining it among @encode/maintainers (myself included).

There are currently 11 open issues and 9 pull requests related to GraphQL. Meanwhile, the last real change to `GraphQLApp` was in October 2019 (https://github.com/encode/starlette/pull/623).

FastAPI does [advertise](https://fastapi.tiangolo.com/advanced/graphql/) GraphQL support via Starlette, but as far as I can tell doesn't actually integrate with it in any way.

This is a proposal and intended just to hopefully get the ball rolling and garner some discussion.

### CODE REVIEW

File: mkdocs.yml
Comment (JayH5) [inline mkdocs.yml:39]: This is to get the nice warning message to render:
Screenshot 2021-02-03 at 22 00 21

**Comment by florimondmanca** [comment]
Sounds like a clear yes to me. There are already several GraphQL specific projects that support ASGI (Strawberry, Ariadne, Tartiflette) (edit: just like you wrote in the docs :)), which makes them trivial to integrate within a Starlette based app (eg via submounting). Definitely better let specialized libraries deal with GraphQL which is an entire world by itself.

**Review by florimondmanca — APPROVED** [review]
Not sure what's left to figure out to get this merge able but what's here looks good to me. :)

**Comment by erewok** [comment]
I agree that this is the right direction to go in and I support this proposal.

**Comment by JayH5** [comment]
Cool, thanks. I guess I had two things in mind before merging this:
1. Possibly refining the warning messages a bit.
2. Making sure the right people know about this.

So for 2):
* @tiangolo hi 👋 if you're around, for if/when FastAPI updates its Starlette dependency, this is something to keep in mind.
* @ciscorn hey 👋 I've linked your project in the docs here. As one of the smaller projects listed I just wanted to make sure that's okay with you? 🙂

**Comment by ciscorn** [comment]
@JayH5 No problem. Thank you! 🙂

**Comment by Kludex** [comment]
I just want to say that I'm happy with this approach and that makes a lot of sense on doing it.

Also, if this approach is followed the same way in the future, it will be cool. I'm saying this because the [removal of UJSONResponse](https://github.com/encode/starlette/pull/1047) which represented a breaking change (even if it's easy to implement).

**Comment by rafalp** [comment]
Do it

**Review by rafalp — APPROVED** [review]

**Comment by JayH5** [comment]
I've tweaked the warning messages so that the Starlette version is mentioned in the docs rather than the code (since the code is already versioned but the docs are not).

Thinking I'll keep #619 open until `GraphQLApp` is actually removed. Will merge this soon unless any last objections 🙂

**Comment by lightoyou** [comment]
@tiangolo can you give recommandation ? Are you going to integrate something inside fast-api ? or let's people choose ?

**Comment by mecampbellsoup** [comment]
> @tiangolo can you give recommandation ? Are you going to integrate something inside fast-api ? or let's people choose ?

https://github.com/encode/starlette/pull/1135#issuecomment-772880130

**Pull request merged**
